pkg/operator/controller/controller_dns_service.go

```diff
@@ -64,3 +71,45 @@ func desiredDNSService(dns *operatorv1.DNS, clusterIP string, daemonsetRef metav
 	}
 	return s
 }
+
+func (r *reconciler) updateDNSService(current, desired *corev1.Service) error {
+	changed, updated := serviceChanged(current, desired)
+	if !changed {
+		return nil
+	}
+
+	if err := r.client.Update(context.TODO(), updated); err != nil {
+		return fmt.Errorf("failed to update dns service %s/%s: %v", updated.Namespace, updated.Name, err)
+	}
+	logrus.Infof("updated dns service: %s/%s", updated.Namespace, updated.Name)
+	return nil
+}
+
+func serviceChanged(current, expected *corev1.Service) (bool, *corev1.Service) {
+	changed := false
+	updated := current.DeepCopy()
+	if !cmp.Equal(current.Spec.TopologyKeys, expected.Spec.TopologyKeys, cmpopts.EquateEmpty()) {
+		updated.Spec.TopologyKeys = expected.Spec.TopologyKeys
+		changed = true
```

Comment:
Is it better to compare fields explicitly, or to use a broad comparison and ignore specific fields?  I have been leaning towards the latter, but the caveat there is that we might get spurious updates if we miss some field that the API sets:  https://github.com/openshift/cluster-ingress-operator/blob/1f3fbf43b4de34f0185271f894372f55252bee43/pkg/operator/controller/ingress/nodeport_service.go#L134

Reply:
Good idea, refactored